### ISSUE TITLE
Wrong built-in precision metric for estimator

### ISSUE THREAD

**Issue by bingyuanlee** [tensorflow/tensorflow#26465]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Y
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):  Ubuntu 16.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): Source
- TensorFlow version (use command below): 1.13.0
- Python version: 3.7
- Bazel version (if compiling from source): N/A
- GCC/Compiler version (if compiling from source): N/A
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A

**Describe the current behavior**
During my experiment with binary classification, I've noticed a significant discrepancy between _tf.estimator.DNNClassifier_ built-in metric of _precision_ from the one computed by _tf.metrics.precision_.

**Describe the expected behavior**
If the calculations were correct. The score of _precision_ should agree with that of _tf.metrics.precision_.

**Code to reproduce the issue**
```
import numpy as np
import tensorflow as tf

# tf.enable_eager_execution()

# Metadata describing the text columns
CSV_COLUMNS = "...".split(',')

DEFAULTS = [...]

KEY_COLUMN = 'KEY'

LABEL_COLUMN = 'LABEL'

# Create an input function reading a file using the Dataset API
# Then provide the results to the Estimator API
def read_dataset(filename, mode, batch_size = 128):
    def _input_fn():
        def decode_csv(value_column):
            columns = tf.decode_csv(value_column, record_defaults=DEFAULTS)
            features = dict(zip(CSV_COLUMNS, columns))
            
            label = features.pop(LABEL_COLUMN)
            
            return features, label

        # Create list of files that match pattern
        file_list = tf.gfile.Glob(filename)

        # Create dataset from file list
        dataset = (tf.data.TextLineDataset(file_list).skip(1)  # Read text file, skip header
                     .map(decode_csv))  # Transform each elem by applying decode_csv fn
        
        if mode == tf.estimator.ModeKeys.TRAIN:
            num_epochs = None # indefinitely
            dataset = dataset.shuffle(buffer_size=10*batch_size)
            dataset = dataset.repeat(num_epochs).batch(batch_size)
        else:
            num_epochs = 1 # end-of-input after this
            dataset = dataset.repeat(num_epochs).batch(10000)
        
        return dataset
    return _input_fn

# Define feature columns
def get_categorical(key, hash_bucket_size):
    return tf.feature_column.categorical_column_with_hash_bucket(key, hash_bucket_size)

def get_numeric(key):
    return tf.feature_column.numeric_column(key)

def get_cols():
  # Define column types
  return [\
             ...
tf.feature_column.indicator_column(tf.feature_column.categorical_column_with_identity('PAXCOUNT',10)),
             ...
      ]

# Create serving input function to be able to serve predictions later using provided inputs
def serving_input_fn():
    feature_placeholders = {
        ...
        'PAXCOUNT': tf.placeholder(tf.int32, [None]),
        ...
        KEY_COLUMN: tf.placeholder_with_default(tf.constant(['nokey']), [None])
    }
    features = {
        key: tf.expand_dims(tensor, -1)
        for key, tensor in feature_placeholders.items()
    }
    return tf.estimator.export.ServingInputReceiver(features, feature_placeholders)

# I've defined additional metric here
def my_precision(labels, predictions):
    return {'my_precision': tf.metrics.precision(labels, 
                                                 tf.strings.to_number(predictions['classes']))}

# I've defined additional metric here
def my_recall(labels, predictions):
    return {'my_recall': tf.metrics.recall(labels,
                                           tf.strings.to_number(predictions['classes']))}

# I've defined additional metric here
def my_f1_score(labels, predictions):
    return {'my_f1_score': tf.contrib.metrics.f1_score(labels, 
                                                    tf.strings.to_number(predictions['classes']))}

estimator = tf.estimator.DNNClassifier(    
                   hidden_units = [256,128,64],
                   feature_columns = get_cols(),
                   model_dir = outdir,
                   n_classes = 2, 
                   optimizer = tf.train.AdamOptimizer(learning_rate=0.001),
                   batch_norm = False)
    
# Custom metric is added here
estimator = tf.estimator.add_metrics(estimator, my_f1_score)
estimator = tf.estimator.add_metrics(estimator, my_precision)
estimator = tf.estimator.add_metrics(estimator, my_recall)

estimator.train(input_fn=read_dataset('train.csv', mode = tf.estimator.ModeKeys.TRAIN), steps = 1000)
estimator.evaluate(input_fn=read_dataset('eval.csv', mode = tf.estimator.ModeKeys.EVAL))
```

**Other info / logs**
**Output of Evaluation**
```
{'accuracy': 0.73864836,
 'accuracy_baseline': 0.5066931,
 'auc': 0.8154568,
 'auc_precision_recall': 0.8016961,
 'average_loss': 0.5324211,
 'label/mean': 0.5066931,
 'loss': 7318.703,
 'my_f1_score': 0.6467407,
 'my_precision': 0.6260541,
 'my_recall': 0.66884124,
 'precision': 0.78366596,
 'prediction/mean': 0.46659818,
 'recall': 0.66884094,
 'global_step': 100}
```
Notice _precision_ is 0.78 while _my_precision_ is 0.62, thats a big difference!

Additionally, I have also added _my_ recall_ and _my_f1 score_, using _tf.metrics.recall_ and _tf.contrib.metrics.f1_score_ respectively.

Taking F1_Score = 2 * ((Precision * Recall) / (Precision + Recall)) to validate F1Score, it seems to agree with _my_precision_ instead of the built-in _precision_ metric.

Does anyone face a similar issue like the above?

**Comment by jvishnuvardhan** [comment]
@bingyuanlee Please fill the issue [template](https://github.com/tensorflow/tensorflow/issues/new?template=10-build-installation-issue.md). Could you update them if they are relevant in your case, or leave them as N/A? Along with the template, please provide as many details as possible to find the root cause of the issue. It would be great if you can provide a small code to reproduce the error. Thanks!

**Comment by bingyuanlee** [comment]
Hi @jvishnuvardhan, I've modified the thread following the template.
Appreciate your help!

**Comment by bingyuanlee** [comment]
Further testing suggests that the discrepancy may or may not happen when there is a change in input.

Though, all three of the metrics that are manually added seem to agree regardless of inputs.

**Comment by karmel** [comment]
@yhliang2018 -- can you comment on the differences here? My guess is that the default aggregation for add_metric is different than for DNNClassifier or something like that?

**Comment by yhliang2018** [comment]
@bingyuanlee the `precision` metric in evaluation is based `labels` and `class_ids`: https://github.com/tensorflow/estimator/blob/afa11287d6d01c6e44e31b54971fc5d1e4712a3d/tensorflow_estimator/python/estimator/canned/head.py#L1043, where `class_ids` is computed from the `logits`: https://github.com/tensorflow/estimator/blob/afa11287d6d01c6e44e31b54971fc5d1e4712a3d/tensorflow_estimator/python/estimator/canned/head.py#L1178; while your `my_precision` is based on `labels` and the `predictions['classes']` (which is from the predictions). That's why they produce different results.

**Comment by bingyuanlee** [comment]
@yhliang2018 : Appreciate the response. I was intending to evaluate the model of its final predicted labels. However, how do we explain the values between `recall` and `my_recall` then, since both evaluate to identical value? Seems like we are facing a standardization issue here..

**Comment by yhliang2018** [comment]
@bingyuanlee Ah, I see the problem. Can you try `tf.contrib.estimator.add_metrics` for your `add_metrics` ? The `tf.estimator.add_metrics` API uses the metrics of TF 2.0, and may produce different results.

**Comment by bingyuanlee** [comment]
@yhliang2018 : Here's another sample result once I used `tf.contrib.estimator.add_metrics` instead. Apparently, this is not the cause of the issue. 

`INFO:tensorflow:Saving dict for global step 1000: accuracy = 0.7981806, accuracy_baseline = 0.5024219, auc = 0.8651023, auc_precision_recall = 0.86191803, average_loss = 0.45824683, global_step = 1000, label/mean = 0.5024219, loss = 5030.3315, my_f1_score = 0.7808682, my_precision = 0.7731018, my_recall = 0.7887923, precision = 0.8054877, prediction/mean = 0.50340647, recall = 0.788792`

**Comment by mohantym** [comment]
Hi @bingyuanlee !We see that you are using old version of Tensorflow which is officially considered as end of life, We recommend that you upgrade to 2.6 version and let us know if the issue still persists in newer versions .Please open a new issue in case you face any errors, we will get you the right help .Thanks!
